sdk/monitor/Azure.Monitor.OpenTelemetry.LiveMetrics/src/Internals/Manager.State.cs

```diff
@@ -0,0 +1,64 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Diagnostics.CodeAnalysis;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Azure.Monitor.OpenTelemetry.LiveMetrics.Internals
+{
+    /// <summary>
+    /// This partial class encapsulates the State.
+    /// This controls if we are in either the Ping or Post state.
+    /// </summary>
+    internal partial class Manager
+    {
+        private Action _callbackAction = () => { };
+
+        internal readonly State _state = new();
+
+        internal TimeSpan _period;
+        private readonly TimeSpan _pingPeriod = TimeSpan.FromSeconds(5);
+        private readonly TimeSpan _postPeriod = TimeSpan.FromSeconds(1);
+        // TODO: WILL NEED ADDITIONAL PERIODS DEFINED FOR BACKOFF.
+
+        private void InitializeState()
+        {
+            SetPingState();
+            Task.Run(() => Run(CancellationToken.None)); // TODO: USE AN ACTUAL CANCELLATION TOKEN
+        }
+
+        private void SetPingState()
+        {
+            _state.Update(LiveMetricsState.Ping);
+            _callbackAction = OnPing;
+            _period = _pingPeriod;
+        }
+
+        private void SetPostState()
+        {
+            _state.Update(LiveMetricsState.Post);
+            _callbackAction = OnPost;
+            _period = _postPeriod;
+        }
+
+        [SuppressMessage("Usage", "AZC0102: Do not use GetAwaiter().GetResult().", Justification = "The EnsureCompleted() extension method is a wrapper around GetAwaiter().GetResult() without adding value. Worse, it breaks during Debug making it impossible to test Live Metrics.")]
+        private void Run(CancellationToken cancellationToken)
+        {
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                var callbackStarted = DateTime.UtcNow;
+
+                _callbackAction.Invoke();
```

Comment:
What would happen, when this call and either of the set is invoked at same time?

Reply:
The scenario you described; I don't think that can happen. 
Invocations are done one at a time, and we wait for the current one to complete before the Task.Delay to wait for the next Invoke. The State (ping/post) is only ever changed inside an Invoke. 

For example:
- Invoke (ping) -> OnPing -> change state to Post -> complete.
- Task.Delay()
- Invoke(post) -> OnPost -> change back to Ping -> complete
- Task.Delay()


Worst case, if an extra Ping is invoked the LiveMetrics service would keep returning the Post command. Our implementation would eventually settle into the Post state. But again, I don't think this would happen. I'll ping you offline to make sure I haven't overlooked something.


For the future, we will have additional States: like Backoff or Filtering that will need to be considered. We could change the State implementation to use Interlocked.Exchange to confirm that a state is changed as an atomic operation.